client/packages/programs/src/JsonForms/components/Search/SearchWithDocumentSource.tsx

```diff
@@ -0,0 +1,142 @@
+import {
+  useTranslation,
+  Box,
+  FormLabel,
+  Typography,
+} from '@openmsupply-client/common';
+import {
+  FORM_LABEL_COLUMN_WIDTH,
+  FORM_INPUT_COLUMN_WIDTH,
+} from '../../common/styleConstants';
+import { z } from 'zod';
+import { useZodOptionsValidation } from '../../common/hooks/useZodOptionsValidation';
+import { usePatientStore } from 'packages/programs/src/hooks';
+import { useEncounter } from '../../../api';
+import { RegexUtils, JSXFormatters } from '@openmsupply-client/common';
+
+const Options = z
+  .object({
+    source: z.enum(['document']),
+    /**
+     * Which pre-defined query to use (in useSearchQueries)
+     */
+    document: z.enum(['patient', 'encounter']),
+    /**
+     * Path in the specified document to extract
+     */
+    path: z.string(),
+    /**
+     * Pattern for formatting selected result (as above)
+     */
+    displayString: z.string().optional(),
+    /**
+     * List of fields to save in document data (from selected item object)
+     */
+    saveFields: z.array(z.string()).optional(),
+  })
+  .strict();
+
+type Options = z.infer<typeof Options>;
+
+const { formatTemplateString } = RegexUtils;
+const { replaceHTMLlineBreaks } = JSXFormatters;
+
+export const SearchWithDocumentSource = (props: ControlProps) => {
+  const {
+    errors: formErrors,
+    path,
+    handleChange,
+    label,
+    uischema,
+    visible,
+  } = props;
+  const { errors: zErrors, options } = useZodOptionsValidation(
+    Options,
+    uischema.options
+  );
+  const { currentPatient = {} } = usePatientStore();
+
+  const {
+    data: currentEncounter,
+    isLoading,
+    isError,
+  } = useEncounter.document.byId(useEncounter.utils.idFromUrl());
+
+  const t = useTranslation('programs');
+
+  const documentData =
+    options?.document === 'patient'
+      ? currentPatient
+      : options?.document === 'encounter'
+      ? currentEncounter
+      : undefined;
+
+  const requestedData = extractProperty(documentData, options?.path ?? '');
+
+  useEffect(() => {
+    if (!requestedData) return;
+
+    if (!options?.saveFields) {
+      handleChange(path, requestedData);
+      return;
+    }
+
+    const newObj: Record<string, any> = {};
+    options.saveFields?.forEach(
+      field => (newObj[field] = requestedData[field] ?? null)
+    );
+    handleChange(path, newObj);
+  }, [requestedData]);
+
+  const displayElement = (
+    <Typography>
+      {options?.displayString ? (
+        replaceHTMLlineBreaks(
+          formatTemplateString(options.displayString, requestedData ?? {}, '')
+        )
+      ) : (
+        <pre>{JSON.stringify(requestedData, null, 2)}</pre>
+      )}
+    </Typography>
+  );
+
+  const error = zErrors
+    ? zErrors
+    : formErrors
+    ? formErrors
+    : isLoading
+    ? null
+    : isError || !documentData
+    ? t('control.search.error.no-document')
+    : !requestedData
+    ? t('control.search.error.no-data', { docPath: options?.path })
+    : null;
```

Comment:
Bit ugly, but hopefully makes sense.

Reply:
this is really hard to read - can you reformat as a method / switch statement / something.
I would go with a switch, that makes each condition clear - it's very hard to quickly follow nested ternary operators like this.

Reply:
Sure, I made a small function with fall-through early returns: ba721005e (Switch didn't really work as there's not a clear single switch expression to compare against)

In general I don't find nested ternaries too hard to follow (they basically just fall through until something is true, then return the next thing), but I agree that this is a particularly gnarly one.

Reply:
🙏  I do find it easier to read